File: packages/cli/src/runner/macos-vm/runner-binary.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";
import { execSync } from "node:child_process";
import { pipeline } from "node:stream/promises";

// Pinned to match the version the DTU pretends to be (see
// packages/dtu-github-actions/src/server/routes/actions/index.ts). Override
// with LOCAL_CI_MACOS_RUNNER_VERSION if you need to test a newer release.
export const DEFAULT_MACOS_RUNNER_VERSION = "2.331.0";

export function resolveMacosRunnerVersion(): string {
  return process.env.LOCAL_CI_MACOS_RUNNER_VERSION?.trim() || DEFAULT_MACOS_RUNNER_VERSION;
}

export function macosRunnerTarballUrl(version: string): string {
  return `https://github.com/actions/runner/releases/download/v${version}/actions-runner-osx-arm64-${version}.tar.gz`;
}

export interface CachedRunner {
  version: string;
  /** Absolute path to the extracted runner directory (contains run.sh etc.). */
  dir: string;
}

// Download + extract the macOS actions-runner tarball for the given version,
// caching the result under <cacheRoot>/<version>/. Subsequent calls for the
// same version return immediately from cache.
export async function ensureMacosRunnerBinary(
  cacheRoot: string,
  version: string = resolveMacosRunnerVersion(),
): Promise<CachedRunner> {
  const dir = path.join(cacheRoot, version);
  const markerFile = path.join(dir, ".extracted");
  const runShExists = fs.existsSync(path.join(dir, "run.sh"));

  if (fs.existsSync(markerFile) && runShExists) {
    return { version, dir };
  }

  await fsp.mkdir(dir, { recursive: true });

  const tarball = path.join(dir, `actions-runner-osx-arm64-${version}.tar.gz`);
  if (!fs.existsSync(tarball)) {
    await downloadToFile(macosRunnerTarballUrl(version), tarball);
  }

  execSync(`tar -xzf ${shellQuote(tarball)} -C ${shellQuote(dir)}`, { stdio: "pipe" });

  // Tarball self-contains run.sh / config.sh / bin/ / externals/. If the shape
  // ever changes we want to notice immediately, not at SSH-exec time.
  if (!fs.existsSync(path.join(dir, "run.sh"))) {
    throw new Error(
      `Extracted runner at ${dir} does not contain run.sh — tarball structure changed?`,
    );
  }

  await fsp.writeFile(markerFile, new Date().toISOString());
  // Keep the tarball around so a corrupt extraction can be recovered without
  // re-downloading. It's ~200MB — the disk cost is negligible next to the
  // 60GB base image.
  return { version, dir };
}

async function downloadToFile(url: string, dst: string): Promise<void> {
  const res = await fetch(url);
  if (!res.ok || !res.body) {
    throw new Error(`Failed to download ${url}: ${res.status} ${res.statusText}`);
  }
  const tmp = dst + ".partial";
  try {
    await pipeline(res.body as unknown as NodeJS.ReadableStream, fs.createWriteStream(tmp));
    await fsp.rename(tmp, dst);
  } catch (err) {
    await fsp.rm(tmp, { force: true }).catch(() => {});
    throw err;
  }
}

function shellQuote(s: string): string {
  return `'${s.replace(/'/g, "'\\''")}'`;
}

File: crates/local-ci-runtime/src/macos_vm/runner_binary.rs
use super::*;

pub fn resolve_macos_runner_version(env_value: Option<&str>) -> String {
    env_value
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .unwrap_or(DEFAULT_MACOS_RUNNER_VERSION)
        .to_owned()
}

pub fn macos_runner_tarball_url(version: &str) -> String {
    format!(
        "https://github.com/actions/runner/releases/download/v{version}/actions-runner-osx-arm64-{version}.tar.gz"
    )
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CachedRunner {
    pub version: String,
    pub dir: PathBuf,
}

pub trait RunnerBinaryIo {
    fn download_to_file(&mut self, url: &str, dst: &Path) -> Result<(), String>;
    fn extract_tarball(&mut self, tarball: &Path, dst: &Path) -> Result<(), String>;
}

#[derive(Debug, Default)]
pub struct CommandRunnerBinaryIo;

impl RunnerBinaryIo for CommandRunnerBinaryIo {
    fn download_to_file(&mut self, url: &str, dst: &Path) -> Result<(), String> {
        let status = Command::new("curl")
            .args(["-fsSL", "-o"])
            .arg(dst)
            .arg(url)
            .status()
            .map_err(|err| format!("failed to run curl: {err}"))?;
        if status.success() {
            Ok(())
        } else {
            Err(format!("curl failed while downloading {url}"))
        }
    }

    fn extract_tarball(&mut self, tarball: &Path, dst: &Path) -> Result<(), String> {
        let status = Command::new("tar")
            .arg("-xzf")
            .arg(tarball)
            .arg("-C")
            .arg(dst)
            .status()
            .map_err(|err| format!("failed to run tar: {err}"))?;
        if status.success() {
            Ok(())
        } else {
            Err(format!("tar failed while extracting {}", tarball.display()))
        }
    }
}

pub fn ensure_macos_runner_binary(
    io: &mut impl RunnerBinaryIo,
    cache_root: &Path,
    version: &str,
) -> Result<CachedRunner, String> {
    let dir = cache_root.join(version);
    let marker = dir.join(".extracted");
    if marker.exists() && dir.join("run.sh").exists() {
        return Ok(CachedRunner {
            version: version.to_owned(),
            dir,
        });
    }

    fs::create_dir_all(&dir).map_err(|err| err.to_string())?;
    let tarball = dir.join(format!("actions-runner-osx-arm64-{version}.tar.gz"));
    if !tarball.exists() {
        io.download_to_file(&macos_runner_tarball_url(version), &tarball)?;
    }
    io.extract_tarball(&tarball, &dir)?;
    if !dir.join("run.sh").exists() {
        return Err(format!(
            "Extracted runner at {} does not contain run.sh — tarball structure changed?",
            dir.display()
        ));
    }
    fs::write(marker, "extracted\n").map_err(|err| err.to_string())?;
    Ok(CachedRunner {
        version: version.to_owned(),
        dir,
    })
}
